### PR TITLE
Fix data race when exceptions happen in SegmentReader.

### LINKED ISSUE #5891: tiflash crash frequently when high currency query
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

high currency query, tiflash crash frequently
```
[2022/09/15 03:53:32.238 +08:00] [ERROR] [BaseDaemon.cpp:377] [BaseDaemon:########################################] [thread_id=5119]
[2022/09/15 03:53:32.239 +08:00] [ERROR] [BaseDaemon.cpp:378] ["BaseDaemon:(from thread 35) Received signal Segmentation fault(11)."] [thread_id=5119]
[2022/09/15 03:53:32.239 +08:00] [ERROR] [BaseDaemon.cpp:408] ["BaseDaemon:Address: 0x248"] [thread_id=5119]
[2022/09/15 03:53:32.239 +08:00] [ERROR] [BaseDaemon.cpp:414] ["BaseDaemon:Access: read."] [thread_id=5119]
[2022/09/15 03:53:32.239 +08:00] [ERROR] [BaseDaemon.cpp:423] ["BaseDaemon:Address not mapped to object."] [thread_id=5119]
[2022/09/15 03:53:32.239 +08:00] [ERROR] [BaseDaemon.cpp:570] ["BaseDaemon:
       0x6703621    faultSignalHandler(int, siginfo_t*, void*) [tiflash+108017185]
                    libs/libdaemon/src/BaseDaemon.cpp:221
  0x7fa5c5c08630    <unknown symbol> [libpthread.so.0+63024]
       0x6efd614    DB::DM::MergedTask::initOnce() [tiflash+116381204]
                    dbms/src/Storages/DeltaMerge/ReadThread/MergedTask.cpp:34
       0x6f03b1d    DB::DM::SegmentReader::run() [tiflash+116407069]
                    dbms/src/Storages/DeltaMerge/ReadThread/SegmentReader.cpp:152
       0x6f05472    void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void (DB::DM::SegmentReader::*)(), DB::DM::SegmentReader*> >(void*) [tiflash+116413554]
                    /usr/local/bin/../include/c++/v1/thread:291
  0x7fa5c5c00ea5    start_thread [libpthread.so.0+32421]"] [thread_id=5119]
```
### 4. What is your TiFlash version? (Required)
master a6bc771f2b3cb1d08c628b649ff9ca796e026932

### PR DESCRIPTION
**Pull request by JinheLin**
### What problem does this PR solve?

Issue Number: close #5891 

Problem Summary:

1. In the issue above, there are a lot of exceptions happened due to memory usage limitation.
2. In `SegmentReader::readSegments`, `merged_task` will be pushed back to `merge_task_pool` when exceptions happened and can be accessed by another read thread if it is scheduled. 
3. However, in this time, read thread can still access to this `merged_task` object and set exception message to it. So data race happened.

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Storages/DeltaMerge/ReadThread/SegmentReader.cpp [hunk 1]
```diff
@@ -116,6 +109,11 @@ class SegmentReader
             {
                 LOG_FMT_DEBUG(log, "All finished, pool_ids={} segment_id={} read_count={}", merged_task->getPoolIds(), merged_task->getSegmentId(), read_count);
             }
+
+            if (!merged_task->allStreamsFinished())
```
Comment (flowbehappy) [inline dbms/src/Storages/DeltaMerge/ReadThread/SegmentReader.cpp:113]: Please add some test cases, you can use fail point to trigger exceptions.

Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/ReadThread/SegmentReader.cpp:113]: Add some comments that why we should not re-push the merged task if exception thrown in the previous lines

Comment (JaySon-Huang) [reply dbms/src/Storages/DeltaMerge/ReadThread/SegmentReader.cpp:113]: And can you remind the reason why you use `SCOPE_EXIT` to re-push the merged task?

**Review by JaySon-Huang — APPROVED** [review]
lgtm

**Review by flowbehappy — APPROVED** [review]

**Pull request merged**